src/System.Data.Common/src/System/Data/Common/AdapterUtil.cs

```diff
@@ -158,9 +160,105 @@ static internal Exception InternalError(InternalErrorCode internalError)
+        static internal Exception InvalidDataLength(long length)
+        {
+            return IndexOutOfRange(SR.Format(SR.SQL_InvalidDataLength, length.ToString(CultureInfo.InvariantCulture)));
+        }
+        static internal IndexOutOfRangeException InvalidBufferSizeOrIndex(int numBytes, int bufferIndex)
+        {
+            return IndexOutOfRange(SR.Format(SR.SQL_InvalidBufferSizeOrIndex, numBytes.ToString(CultureInfo.InvariantCulture), bufferIndex.ToString(CultureInfo.InvariantCulture)));
+        }
+        static internal ArgumentOutOfRangeException InvalidDestinationBufferIndex(int maxLen, int dstOffset, string parameterName)
+        {
+            return ArgumentOutOfRange(SR.Format(SR.ADP_InvalidDestinationBufferIndex, maxLen.ToString(CultureInfo.InvariantCulture), dstOffset.ToString(CultureInfo.InvariantCulture)), parameterName);
+        }
+
+
+        // { "a", "a", "a" } -> { "a", "a1", "a2" }
+        // { "a", "a", "a1" } -> { "a", "a2", "a1" }
+        // { "a", "A", "a" } -> { "a", "A1", "a2" }
+        // { "a", "A", "a1" } -> { "a", "A2", "a1" }
+        static internal void BuildSchemaTableInfoTableNames(string[] columnNameArray)
+        {
+            Dictionary<string, int> hash = new Dictionary<string, int>(columnNameArray.Length);
+
+            int startIndex = columnNameArray.Length; // lowest non-unique index
+            for (int i = columnNameArray.Length - 1; 0 <= i; --i)
+            {
+                string columnName = columnNameArray[i];
+                if ((null != columnName) && (0 < columnName.Length))
+                {
+                    columnName = columnName.ToLowerInvariant();
+                    int index;
+                    if (hash.TryGetValue(columnName, out index))
+                    {
+                        startIndex = Math.Min(startIndex, index);
+                    }
+                    hash[columnName] = i;
+                }
+                else
+                {
+                    columnNameArray[i] = string.Empty;
+                    startIndex = i;
+                }
+            }
+            int uniqueIndex = 1;
+            for (int i = startIndex; i < columnNameArray.Length; ++i)
+            {
+                string columnName = columnNameArray[i];
+                if (0 == columnName.Length)
+                { // generate a unique name
+                    columnNameArray[i] = "Column";
+                    uniqueIndex = GenerateUniqueName(hash, ref columnNameArray[i], i, uniqueIndex);
+                }
+                else
+                {
+                    columnName = columnName.ToLowerInvariant();
+                    if (i != hash[columnName])
+                    {
+                        GenerateUniqueName(hash, ref columnNameArray[i], i, 1);
+                    }
+                }
+            }
+        }
+        static private int GenerateUniqueName(Dictionary<string, int> hash, ref string columnName, int index, int uniqueIndex)
```

Comment:
Nit: Add a line above this method as a separator between methods? Applies throughout this file.